### PR TITLE
refactor: declaration files (and tests); `connect(): Realm`

### PR DESCRIPTION
**Pull request by cyjake**
changed from `connect(): Bone` to `connect(): Realm` which makes more sense. This behavior is not actually used in any dependents hence is planed with a minor release

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -121,15 +121,25 @@ interface WhereConditions {
...
-  columnType: string,
-  isNullable: boolean,
-  type: boolean | number | string | Date | JSON;
+  column: string;
+  columnType: string;
+  allowNull: boolean;
+  defaultValue: Literal;
+  primaryKey: boolean;
+  dataType: string;
+  jsType: boolean | number | string | Date | JSON;
```
Comment (JimmyDaddy) [inline types/index.d.ts:141]: `Buffer` should be added

File: types/index.d.ts [hunk 2]
```diff
@@ -543,6 +553,27 @@ interface ConnectOptions {
   models: string | Bone[];
 }
 
+interface InitOptions {
+  underscored?: boolean;
+  tableName?: string;
+  hooks?: Record<string, Function>;
```
Comment (JimmyDaddy) [inline types/index.d.ts:559]: `hooks` should be `Record<string, AsyncFunction>`

File: types/index.d.ts [hunk 3]
```diff
@@ -103,7 +103,7 @@ declare class Spell {
   toString(): string;
 }
 
-type Literal = null | boolean | number | string | Date
+type Literal = null | boolean | number | string | Date | JSON | ArrayBuffer;
```
Comment (cyjake) [inline types/index.d.ts:106]: Buffer requires `@types/node`, added ArrayBuffer instead...

**Review by JimmyDaddy — APPROVED** [review]

File: types/index.d.ts [hunk 4]
```diff
@@ -543,6 +553,31 @@ interface ConnectOptions {
...
 }
 
+interface InitOptions {
+  underscored?: boolean;
+  tableName?: string;
+  hooks?: {
+    [key in 'beforeCreate' | 'beforeBulkCreate' | 'beforeUpdate' | 'beforeSave' |  'beforeUpsert' | 'beforeRemove' ]: (options: QueryOptions) => Promise<void>
+  } | {
+    [key in 'afterCreate' | 'afterBulkCreate' | 'afterUpdate' | 'afterSave' | 'afterUpsert' | 'afterRemove' ]: (instance: Bone, result: Object) => Promise<void>
+  };
```
Comment (cyjake) [inline types/index.d.ts:563]: added before/after names

It seems typescript removed `AsyncFunction`, added specific function type definitions. Static hooks not added yet.

**Pull request merged**